higher/index.ts

```diff
@@ -0,0 +1,2314 @@
+/*
+ **************
+ **************
+ *** Higher ***
+ **************
+ **************
+ */
+
+// TODO thoughts: should permission checks happen at the
+// database layer?
+// e.g. any local database modifications should either be
+// checked (it is possible they come from an external party)
+// or unchecked (it is impossible for them to come from an
+// external party)
+// how hard is reasoning about this?
+
+// TODO add permission checks for setting cryptographic keys?
+
+// TODO instead of using this.core.olmWrapper.getIdkey() everywhere, idkey is not 
+// expected to change so just set some object state once a device is
+// created (could still be in olmWrapper)
+
+// TODO need new special name for LINKED group (confusing when linking non-LINKED groups)
+
+import EventEmitter from "events";
+import { Core } from "../core/client";
+import { LocalStorageWrapper } from "./db/localStorageWrapper.js";
+
+/* doubly-linked tree, allows cycles */
+const NAME         : string = "name";
+const CONTACT_LEVEL: string = "contactLevel";
+const PARENTS      : string = "parents";
+const CHILDREN     : string = "children";
+const ADMINS       : string = "admins";
+const WRITERS      : string = "writers";
+
+type groupObjType = {
+  id: string,
+  value: groupValType,
+};
+
+type groupValType = {
```

Comment:
This seems way too complicated for what we're trying to do. I think the design of having an application developer set permissions level was a bit better. So for every object, you would have some set of groups, each with its own permissions, rather than different permissions within a group. The change could simplify a lot of this logic so that groups are just lists of users Ids + other groups without all the additional logic for admins/writers/etc...